File: Lex/Lex.java
// ---------------------------------------------------------------------------
// mborisad
// Mayowa Borisade
// CMPS101
// Pa1
// 04/09/18
// Lex.java
// This program creates List that sorts an String array in 
// Lexicological order based off its indices.
// ---------------------------------------------------------------------------

import java.io.*;
import java.util.Scanner;


class Lex{
    public static void main(String[] args) throws IOException{
        Scanner in = null;
        PrintWriter out = null;
        String line = null;
        String[] token = null;
        String[] words;
        int n, lineNumber = 0;
      
      
        if(args.length != 2 ){
            System.err.println("Usage: FileIO infile outfile");
            System.exit(1);
        }
      
        in = new Scanner(new File(args[0]));
        out = new PrintWriter(new FileWriter(args[1]));
        
        
        while(in.hasNextLine()){
            lineNumber++;
            line = in.nextLine()+" ";  
        }
        
        in.close();        
        in = new Scanner(new File(args[0]));
        
        words = new String[lineNumber];
      
        for(int i = 0; i < lineNumber; i++){
            words[i] = in.nextLine();
        }
        
        
        
        
        // This is to put the indices in the write order into the List
        List A = new List();
        
        int c, cp;
        if((words[0]).compareTo(words[1])<0){
            A.append(0);
            A.append(1);
        }else{
            A.append(1);
            A.append(0);
        }
        
        for(int i = 2; i < words.length; i++){
            for(A.moveBack(); A.index() > 0; A.movePrev()){
                c = A.get();
                A.movePrev();
                cp = A.get();
                A.moveNext();
                                
                if((words[c]).compareTo(words[i])==0){
                    A.insertAfter(i);

                }else if(((words[c]).compareTo(words[i])>0) && ((words[cp]).compareTo(words[i]) < 0)){
                    A.insertBefore(i);
                    A.movePrev();
                }
            }
            if(((words[A.back()]).compareTo(words[i]) < 0)){
                A.append(i);
            }else if(((words[A.front()]).compareTo(words[i]) > 0)){
                A.prepend(i);
            }

        }
        
        
        // To print the list in order into the output file 
        for(A.moveFront(); A.index()>=0; A.moveNext()){
            out.println(words[A.get()]);
        }
            

        in.close();
        out.close();      
   
   }

}

File: LexInC/Lex.c
// ---------------------------------------------------------------------------
// mborisad
// Mayowa Borisade
// CMPS101
// Pa2
// 04/23/18
// Lex.c
// This program creates List that sorts an String array in 
// Lexicological order based off its indices.
// ---------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"List.h"
#define MAX_LEN 1100


int main(int argc, char * argv[]){
    int i, count=0;
    FILE *in, *out;
    char line[MAX_LEN];
    
    
    // check command line for correct number of arguments
    if( argc != 3 ){
        printf("Usage: %s <input file> <output file>\n", argv[0]);
        exit(1);
    }

    // open files for reading and writing 
    in = fopen(argv[1], "r");
    out = fopen(argv[2], "w");
    if( in==NULL ){
        printf("Unable to open file %s for reading\n", argv[1]);
        exit(1);
    }
    if( out==NULL ){
        printf("Unable to open file %s for writing\n", argv[2]);
        exit(1);
    }
   
   
   
    /* read each line of input file, then count and print tokens */
    while( fgets(line, MAX_LEN, in) != NULL)  {
        count = count + 1;
    }
    
    fclose(in);
   
    //Declaration of string array   
    char* words[MAX_LEN];
   
   
    // open files for reading and writing 
    in = fopen(argv[1], "r");
   
    if( in==NULL ){
        printf("Unable to open file %s for reading\n", argv[1]);
        exit(1);
    }
   
   
    i = 0;
    /* read each line of input file, then count and print tokens */
    // Not sure if this works
    while(fgets(line, MAX_LEN, in) != NULL)  {
        words[i] = calloc(MAX_LEN, sizeof(char)); 
        strcpy(words[i], line);
        i++;
    }
   
   
   
    // Start ----------------------------------------------------
    // This is to put the indices in the write order into the List
    List A = newList();
        
    int c = -1;
    int a = -1;
	int cp = -1;
	int b = -1;
	int f = -1;
	int g = -1;
        
    for(int k = 0; k < count; k++){
        for(moveBack(A); index(A) > 0; movePrev(A)){
            c = get(A);
            movePrev(A);
            cp = get(A);
            moveNext(A);
                 
            a = strcmp(words[c],words[k]);
            b = strcmp((words[cp]), (words[k]));  
            f = strcmp((words[back(A)]), (words[k]));
            g = strcmp((words[front(A)]), (words[k]));  
            if(a == 0){
                insertAfter(A, k);
            }else if(a > 0 && b < 0){
                insertBefore(A, k);
                movePrev(A);
            }
        }
        if(f < 0){
            append(A, k);
        }else if(g > 0){
            prepend(A, k);
        }
    }
            
    // To print the list in order into the output file 
    for(moveFront(A); index(A)>= 0; moveNext(A)){
        fprintf(out, "%s", words[get(A)]);
    }
        
    
    //To free array
    for(int j = 0; j < count; j++){
        free(words[j]);
        words[j] = NULL;
    }
    
    freeList(&A);
    fclose(in);
    fclose(out);      


}
